src/new_zonefile/simple.rs

```diff
@@ -0,0 +1,686 @@
+        *last_ttl = Some(ttl);
+
+        // Parse the record data.
+        let data =
+            // NOTE: If the unknown record data format is used for a type we
+            // recognize, we will scan it as 'UnknownRecordData' and then
+            // parse the bytes to ensure validity.  Unfortunately, it is not
+            // possible to parse bytes into '&RevName' (it would need access
+            // to the bump allocator).  So we just parse as 'RevNameBuf' and
+            // push all names onto the bump allocator afterwards.  Hopefully,
+            // the compiler is able to inline this fairly large type away.
+            <RecordData<'a, RevNameBuf>>::scan(&mut scanner, alloc, buffer)
+                .map_err(RecordError::DataError)?
+                .map_names(|name| {
+                    let bytes = alloc.alloc_slice_copy(name.as_bytes());
+                    // SAFETY: 'RevName::as_bytes()' is always valid.
+                    unsafe { RevName::from_bytes_unchecked(bytes) }
+                });
+        Ok(Record {
+            rname: owner,
+            rtype: data.rtype(),
+            rclass,
+            ttl,
+            rdata: data,
+        })
+    }
+}
+
+//----------- Type Aliases ---------------------------------------------------
+
+/// A scanned record.
+pub type ScannedRecord<'a> = Record<&'a RevName, RecordData<'a, &'a RevName>>;
+
+/// An entry in a zonefile.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Entry<'a> {
+    /// A zonefile record.
+    Record(ScannedRecord<'a>),
+
+    /// An include directive.
+    Include {
+        /// The file name to include.
+        file_name: PathBuf,
+
+        /// The origin domain name to use.
+        origin: Option<&'a RevName>,
+    },
+}
+
+//----------- ScanError ------------------------------------------------------
+
+/// An error in scanning a zonefile.
+#[derive(Debug)]
+pub enum ZonefileError {
+    /// An error in scanning a record.
+    Record {
+        /// The scanning error that occurred.
+        error: RecordError,
+
+        /// The line number of the entry.
+        line_number: usize,
+    },
+
+    /// An error in scanning a directive.
+    Directive {
+        /// The scanning error that occurred.
+        error: DirectiveError,
+
+        /// The line number of the entry.
+        line_number: usize,
+    },
+
+    /// An error in scanning an entry.
+    Entry {
+        /// The scanning error that occurred.
+        error: ScanError,
+
+        /// The line number of the entry.
+        line_number: usize,
+    },
+
+    /// An error in disambiguating entries.
+    Entries(EntriesError),
+}
+
+//--- Conversion from error types
+
+impl From<EntriesError> for ZonefileError {
+    fn from(error: EntriesError) -> Self {
+        Self::Entries(error)
+    }
+}
+
+//----------- RecordError ----------------------------------------------------
+
+/// An error in scanning a zonefile record.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum RecordError {
```

Comment:
Do we want to leave the error types as (open) enums or rather wrap them away into an opaque type? I don’t think a user would want to match on the variants here?